docs/src/tutorial-nondeterministic-variables.md

```diff
@@ -1,95 +1,114 @@
+2. `quantity`, however, has type `NonZeroU32`.
+In Rust, it would be undefined behavior to have a value of `0` for this type.
 
-You can try it out by running the example under
-[`arbitrary-variables`](https://github.com/model-checking/kani/tree/main/docs/src/tutorial/arbitrary-variables/):
+We included an extra assertion that the value returned by `kani::any()` here was actually non-zero.
+If we run this, you'll notice that verification succeeds.
 
-```
+```bash
 cargo kani --harness safe_update
 ```
 
-## Unsafe nondeterministic variables
+`kani::any()` is safe Rust, and so Kani only implements it for types where type invariants are enforced.
+For `NonZeroU32`, this means we never return a `0` value.
+The assertion we wrote in this harness was just an extra check we added to demonstrate this fact.
 
-Kani also includes an **unsafe** method to generate unconstrained nondeterministic variables which do not take type invariants into consideration.
-As with any unsafe method in Rust, users have to make sure the right guardrails are put in place to avoid undesirable behavior.
+## Bounding nondeterministic variables
 
-That said, there may be cases where you want to verify your code taking into consideration that some inputs may contain invalid data.
+There is no implementation of `kani::any()` for more complex data structures like `Vec`, however.
+The trouble with a nondeterministic vector is that you usually need to _bound_ the size of the vector, for the reasons we investigated in the [last chapter](./tutorial-loop-unwinding.md).
+The `kani::any()` function does not have any arguments, and so cannot be given an upper bound.
 
-Let's see what happens if we modify our verification harness to use the unsafe method `kani::any_raw()` to generate the updated value.
+You can use `kani::any()` for `[T; N]` (if implemented for `T`).
+But this the type of an exactly-sized array.
+Again, `kani::any()` does not have an implementation for an unsized slice (like `[T]`).
 
-```rust
-{{#include tutorial/arbitrary-variables/src/inventory.rs:unsafe_update}}
-```
+This does not mean you cannot have a nondeterministic vector.
+It just means you have to construct one.
+Our example proof harness above constructs a nondeterministic `Inventory` of size `1`, simply by starting with the empty `Inventory` and inserting a nondeterministic entry.
 
-We commented out the assertion that the underlying `u32` variable cannot be `0`, since this no longer holds.
-The verification will now fail showing that the `inventory.get(&id).unwrap()` method call can panic.
+## Custom nondeterministic types
 
-This is an interesting issue that emerges from how `rustc` optimizes the memory layout of `Option<NonZeroU32>`.
-The compiler is able to represent `Option<NonZeroU32>` using `32` bits by using the value `0` to represent `None`.
+There's nothing special about `kani::any()` except that Kani ships with implementations for a few types where we can guarantee safety.
```

Comment:
maybe rephrase this to something like:
> `kani::any()` is the only method that Kani provides to inject non-determinism in a harness, but Kani ships with implementations for just a few types where it can guarantee safety.